File: java/gov/nasa/gsfc/gmsec/api/Status.java
/*
 * Copyright 2007-2021 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file Status.java
 */

package gov.nasa.gsfc.gmsec.api;


/**
 * This is the class for holding status information returned by the API
 * function calls to provide feedback. Status provides an error code
 * and error message specific to the error type, but independent of error
 * source.
 */
public class Status
{
	private StatusClassification m_class;
	private StatusCode           m_code;
	private String               m_reason;
	private int                  m_customCode;


	/**
	 * Constructor.
	 */
	public Status()
	{
		this(StatusClassification.NO_ERROR, StatusCode.NO_ERROR_CODE, "", 0);
	}


	/**
	 * Constructs a Status object with the given error class, code and message.
	 *
	 * @param clazz  Value indicating source of the error
	 * @param code   Value indicating reason for the error
	 * @param reason Explanation for the error
	 *
	 * @see StatusClassification
	 * @see StatusCode
	 */
	public Status(StatusClassification clazz, StatusCode code, String reason)
	{
		this(clazz, code, reason, 0);
	}


	/**
	 * Constructs a Status object with the given error class, code, custom code and message.
	 *
	 * @param clazz      Value indicating source of the error
	 * @param code       Value indicating reason for the error
	 * @param reason     Explanation for the error
	 * @param customCode Custom error code (typically middleware specific) regarding the root cause of the error
	 *
	 * @see StatusClassification
	 * @see StatusCode
	 */
	public Status(StatusClassification clazz, StatusCode code, String reason, int customCode)
	{
		m_class      = clazz;
		m_code       = code;
		m_reason     = reason;
		m_customCode = customCode;
	}


	/**
	 * Constructs a Status object with info from the given GMSEC_Exception object.
	 *
	 * @param e GMSEC_Exception object indicating source/reason for the error
	 *
	 * @throws IllegalArgumentException Thrown if the exception object is null.
	 */
	public Status(GMSEC_Exception e)
		throws IllegalArgumentException
	{
		if (e == null)
		{
			throw new IllegalArgumentException("GMSEC_Exception object is null");
		}

		m_class      = e.getErrorClassification();
		m_code       = e.getErrorCode();
		m_reason     = e.getErrorMessage();
		m_customCode = e.getCustomCode();
	}


	/**
	 * Construct Status object using values from an existing Status object.
	 *
	 * @param other Status object to copy.
	 */
	public Status(Status other)
	{
		if (other != null)
		{
			m_class      = other.m_class;
			m_code       = other.m_code;
			m_reason     = other.m_reason;
			m_customCode = other.m_customCode;
		}
	}


	/**
	 * Used to determine whether the Status object is reporting an error or not.
	 * <p>
	 * Note: Only the status class is examine; the status code is not checked.
	 *
	 * @return False if the status class is set to NO_ERROR; true otherwise.
	 */
	public boolean isError()
	{
		return m_class != StatusClassification.NO_ERROR;
	}


	/**
	 * This function will return a verbose error string that contains the
	 * Status class, code, custom code and reason.
	 * <p>
	 * The format is as follows:
	 * <pre>{@code
	 * [class,code,custom code] : reason
	 * }</pre>
	 *
	 * @return The error string associated with the Status object.
	 */
	public String get()
	{
		StringBuilder sb = new StringBuilder();
		sb.append("[").append(m_class.getValue()).append(",").append(m_code.getValue()).append(",")
			.append(m_customCode).append("] : ").append(m_reason);

		return sb.toString();
	}


	/**
	 * Used to set error class, error code, and error text for the Status object.
	 *
	 * @param clazz  Error class
	 * @param code   Error code
	 * @param reason Error string
	 *
	 * @see StatusClassification
	 * @see StatusCode
	 */
	public void set(StatusClassification clazz, StatusCode code, String reason)
	{
		set(clazz, code, reason, 0);
	}


	/**
	 * Used to set error class, error code, custom code, and error text for the Status object.
	 *
	 * @param clazz      Error class
	 * @param code       Error code
	 * @param reason     Error string
	 * @param customCode Custom code (typically from middleware)
	 *
	 * @see StatusClassification
	 * @see StatusCode
	 */
	public void set(StatusClassification clazz, StatusCode code, String reason, int customCode)
	{
		m_class      = clazz;
		m_code       = code;
		m_reason     = (reason == null ? "" : reason);
		m_customCode = customCode;
	}


	/**
	 * Returns the error class associated with the Status object.
	 *
	 * @return A StatusClassification enumerated value.
	 *
	 * @see StatusClassification
	 */
	public StatusClassification getClassification()
	{
		return m_class;
	}


	/**
	 * This will set the status classification.
	 *
	 * @param clazz Status classification value
	 *
	 * @see StatusClassification
	 */
	public void setClassification(StatusClassification clazz)
	{
		m_class = clazz;
	}


	/**
	 * Returns the status code associated with the Status object.
	 *
	 * @return A StatusCode enumerated value.
	 *
	 * @see StatusCode
	 */
	public StatusCode getCode()
	{
		return m_code;
	}


	/**
	 * This will set the status code.
	 *
	 * @param code Status code enumerated value
	 *
	 * @see StatusCode
	 */
	public void setCode(StatusCode code)
	{
		m_code = code;
	}


	/**
	 * Returns the error message associated with the Status object.
	 *
	 * @return A string containing the error message.
	 */
	public String getReason()
	{
		return m_reason;
	}


	/**
	 * This will set the error string text.
	 *
	 * @param reason The error text string.
	 */
	public void setReason(String reason)
	{
		m_reason = reason;
	}


	/**
	 * Returns the custom code associated with the Status object.
	 *
	 * @return The custom code.
	 */
	public int getCustomCode()
	{
		return m_customCode;
	}


	/**
	 * This will set the (middleware) specific error code.
	 *
	 * @param customCode Error value, typically middleware specific
	 */
	public void setCustomCode(int customCode)
	{
		m_customCode = customCode;
	}


	/**
	 * This clears the Status object to a 'no error' state.
	 */
	public void reset()
	{
		m_class      = StatusClassification.NO_ERROR;
		m_code       = StatusCode.NO_ERROR_CODE;
		m_reason     = "";
		m_customCode = 0;
	}
}

File: framework/src/gmsec5/c/status.cpp
/*
 * Copyright 2007-2023 United States Government as represented by the
 * Administrator of The National Aeronautics and Space Administration.
 * No copyright is claimed in the United States under Title 17, U.S. Code.
 * All Rights Reserved.
 */


/**
 * @file status.cpp
 *
 * @brief This file contains functions for the management of status.
 */


#include <gmsec5/c/status.h>

#include <gmsec5/Status.h>

using namespace gmsec::api5;


GMSEC_Status CALL_TYPE statusCreate(void)
{
	return reinterpret_cast<GMSEC_Status>(new Status());
}


GMSEC_Status CALL_TYPE statusCreateWithValues(GMSEC_I32 clazz, GMSEC_I32 code, const char* reason, GMSEC_I32 custom)
{
	return reinterpret_cast<GMSEC_Status>(new Status(clazz, code, reason, custom));
}


GMSEC_Status CALL_TYPE statusCreateCopy(GMSEC_Status other)
{
	GMSEC_Status created = NULL;
	Status*      status  = reinterpret_cast<Status*>(other);

	if (status)
	{
		created = reinterpret_cast<GMSEC_Status>(new Status(*status));
	}

	return created;
}


void CALL_TYPE statusDestroy(GMSEC_Status* status)
{
	if (status && *status)
	{
		Status* s = reinterpret_cast<Status*>(*status);

		delete s;

		*status = NULL;
	}
}


GMSEC_BOOL CALL_TYPE statusHasError(GMSEC_Status status)
{
	GMSEC_BOOL hasError = GMSEC_BOOL::GMSEC_FALSE;
	Status*    s        = reinterpret_cast<Status*>(status);

	if (s)
	{
		hasError = (s->hasError() ? GMSEC_BOOL::GMSEC_TRUE : GMSEC_BOOL::GMSEC_FALSE);
	}

	return hasError;
}


const char* CALL_TYPE statusGet(GMSEC_Status status)
{
	const char* get = NULL;
	Status*     s   = reinterpret_cast<Status*>(status);

	if (s)
	{
		get = s->get();
	}

	return get;
}


void CALL_TYPE statusSet(GMSEC_Status status, GMSEC_I32 clazz, GMSEC_I32 code, const char* reason, GMSEC_I32 custom)
{
	Status* s = reinterpret_cast<Status*>(status);

	if (s)
	{
		s->set(clazz, code, reason, custom);
	}
}


GMSEC_I32 CALL_TYPE statusGetClass(GMSEC_Status status)
{
	GMSEC_I32 clazz = GMSEC_NO_ERROR;
	Status*   s     = reinterpret_cast<Status*>(status);

	if (s)
	{
		clazz = s->getClass();
	}

	return clazz;
}


void CALL_TYPE statusSetClass(GMSEC_Status status, GMSEC_I32 clazz)
{
	Status* s = reinterpret_cast<Status*>(status);

	if (s)
	{
		s->setClass(clazz);
	}
}


GMSEC_I32 CALL_TYPE statusGetCode(GMSEC_Status status)
{
	GMSEC_I32 code = GMSEC_NO_ERROR_CODE;
	Status*   s    = reinterpret_cast<Status*>(status);

	if (s)
	{
		code = s->getCode();
	}

	return code;
}


void CALL_TYPE statusSetCode(GMSEC_Status status, GMSEC_I32 code)
{
	Status* s = reinterpret_cast<Status*>(status);

	if (s)
	{
		s->setCode(code);
	}
}


const char* CALL_TYPE statusGetReason(GMSEC_Status status)
{
	const char* reason = NULL;
	Status*     s      = reinterpret_cast<Status*>(status);

	if (s)
	{
		reason = s->getReason();
	}

	return reason;
}


void CALL_TYPE statusSetReason(GMSEC_Status status, const char* reason)
{
	Status* s = reinterpret_cast<Status*>(status);

	if (s)
	{
		s->setReason(reason);
	}
}


GMSEC_I32 CALL_TYPE statusGetCustomCode(GMSEC_Status status)
{
	GMSEC_I32 custom = 0;
	Status*   s      = reinterpret_cast<Status*>(status);

	if (s)
	{
		custom = s->getCustomCode();
	}

	return custom;
}


void CALL_TYPE statusSetCustomCode(GMSEC_Status status, GMSEC_I32 custom)
{
	Status* s = reinterpret_cast<Status*>(status);

	if (s)
	{
		s->setCustomCode(custom);
	}
}


void CALL_TYPE statusReset(GMSEC_Status status)
{
	Status* s = reinterpret_cast<Status*>(status);

	if (s)
	{
		s->reset();
	}
}
